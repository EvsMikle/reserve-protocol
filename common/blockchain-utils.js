
const { BigNumber } = require("ethers");

// getChainId: Returns current chain Id
async function getChainId() {
  let _chainId;
  try {
    _chainId = await this.network.provider.send('eth_chainId');
  } catch (e) {
    console.log('failed to get chainId, falling back on net_version...');
    _chainId = await this.network.provider.send('net_version');
  }

  if (!_chainId) {
    throw new Error(`could not get chainId from network`);
  }
  if (_chainId.startsWith('0x')) {
    _chainId = BigNumber.from(_chainId).toString();
  }
  return _chainId;
}

module.exports = {
  getChainId
};
