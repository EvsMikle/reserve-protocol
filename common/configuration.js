const { ZERO_ADDRESS } = require("../common/constants");

const networkConfig = {
    default: {
        name: 'hardhat',
    },
    31337: {
        name: 'localhost',
    },
    3: {
        name: 'ropsten',
        rsrPrev: '0x58408daf0664dc9ff4645414ce5f9ace059f0470',
        compoundMath: '0x2C3E02b6A137F3ac194952B7ea3763631984E49B',
        slowWallet: '0x9d2e46f086be2f76d576c95409344550a6ffd24a',
        multisigWallet: '0xA7b123D54BcEc14b4206dAb796982a6d5aaA6770',
        USDC: '0x07865c6e87b9f70255377e024ace6630c1eaa37f',
    },
    1: {
        name: 'mainnet',
        rsrPrev: '0x8762db106b2c2a0bccb3a80d1ed41273552616e8',
        slowWallet: '0x82734Ae1E495d6e67bF12F2153659a0F74d2874B',
        multisigWallet: '0xA7b123D54BcEc14b4206dAb796982a6d5aaA6770',
        USDC: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'
    }
}

const developmentChains = ["hardhat", "localhost"]

const rTokenConfig = {
    default: {
        name: "RToken",
        symbol: "RTKN",
        params: {
            stakingDepositDelay: 0,
            stakingWithdrawalDelay: 0,
            maxSupply: 0,
            minMintingSize: 0,
            issuanceRate: 0,
            rebalancingFreezeCost: 0,
            insurancePaymentPeriod: 0,
            expansionPerSecond: 0,
            expenditureFactor: 0,
            spread: 0,
            exchange: ZERO_ADDRESS,
            circuitBreaker: ZERO_ADDRESS,
            txFeeCalculator: ZERO_ADDRESS,
            insurancePool: ZERO_ADDRESS,
            protocolFund: ZERO_ADDRESS
        },
        basketTokens: [
            {
                tokenAddress: ZERO_ADDRESS,
                genesisQuantity: 0,
                rateLimit: 1,
                maxTrade: 1,
                priceInRToken: 0,
                slippageTolerance: 0
            }
        ],
        rsr: {
            tokenAddress: ZERO_ADDRESS,
            genesisQuantity: 0,
            rateLimit: 1,
            maxTrade: 1,
            priceInRToken: 0,
            slippageTolerance: 0
        }
    }
}

const getRTokenConfig = (name) => {
    if (!rTokenConfig[name]) {
        throw new Error(`Configuration ${name} not available`);
    }
    return rTokenConfig[name];
}

module.exports = {
    networkConfig,
    developmentChains,
    getRTokenConfig
}
