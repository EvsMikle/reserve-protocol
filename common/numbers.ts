import { BigNumber, BigNumberish } from 'ethers'
import { ethers } from 'hardhat'

import { BN_SCALE_FACTOR, SCALE_DECIMALS } from './constants'

export const ZERO = BigNumber.from(0)

// Convenience form for "BigNumber.from" that also accepts scientific notation
export const bn = (x: BigNumberish): BigNumber => {
  if (typeof x === 'string') return _parseScientific(x)
  return BigNumber.from(x)
}

export const pow10 = (exponent: BigNumberish): BigNumber => {
  return BigNumber.from(10).pow(exponent)
}

// Convert to the BigNumber representing a Fix from a BigNumberish.
// Try to handle fractional values intelligently. In particular:
//     If the arg is a fractional JS number, it will be rounded to 9 decimal places (!) and used that way
//     If the arg is a string, it can be in decimal or scientific notation, and will be handled appropriately

export const fp = (x: BigNumberish): BigNumber => {
  if (typeof x === 'string') return _parseScientific(x, SCALE_DECIMALS)
  if (typeof x === 'number' && !Number.isInteger(x))
    return _parseScientific(x.toFixed(9), SCALE_DECIMALS)
  return BigNumber.from(x).mul(pow10(SCALE_DECIMALS))
}

export const divCeil = (x: BigNumber, y: BigNumber): BigNumber =>
  // ceil(x/y) == (x + y - 1) / y
  x.add(y).sub(1).div(y)

// Wheter the absolute difference between x and y is less than z
export const near = (x: BigNumber, y: BigNumber, z: BigNumberish): boolean => {
  if (x.lt(y)) {
    return y.sub(x).lte(z)
  }
  return x.sub(y).lte(z)
}

// _parseScientific(s, scale) returns a BigNumber with value (s * 10**scale),
// where s is a string in decimal or scientific notation,
// and scale is a BigNumberish indicating a number of additional zeroes to add to the right,
// Fractional digits in the result are truncated.
// TODO: Maybe we should error if we're truncating digits instead?
//
// A few examples:
//     _parseScientific('1.4e2') == BigNumber.from(140)
//     _parseScientific('-2') == BigNumber.from(-2)
//     _parseScientific('0.5', 18) == BigNumber.from(5).mul(pow10(17))
//     _parseScientific('0.127e2') == BigNumber.from(12)
function _parseScientific(s: string, scale: BigNumberish = 0): BigNumber {
  // Scientific Notation: <INT>(.<DIGITS>)?(e<INT>)?
  // INT: [+-]?DIGITS
  // DIGITS: \d+
  const match = s.match(
    /^(?<sign>[+-]?)(?<int_part>\d+)(\.(?<frac_part>\d+))?(e(?<exponent>[+-]?\d+))?$/
  )
  if (!match || !match.groups) throw new Error(`Illegal decimal string ${s}`)

  let sign = match.groups.sign === '-' ? -1 : 1
  let int_part = BigNumber.from(match.groups.int_part)
  const frac_part = match.groups.frac_part ? BigNumber.from(match.groups.frac_part) : ZERO
  let exponent = match.groups.exponent ? BigNumber.from(match.groups.exponent) : ZERO
  exponent = exponent.add(scale)

  // "zero" the fractional part by shifting it into int_part, keeping the overall value equal
  if (!frac_part.eq(ZERO)) {
    const shift_digits = match.groups.frac_part.length
    int_part = int_part.mul(pow10(shift_digits)).add(frac_part)
    exponent = exponent.sub(shift_digits)
  }

  // Shift int_part left or right as exponent requires
  const positive_output: BigNumber = exponent.gte(ZERO)
    ? int_part.mul(pow10(exponent))
    : int_part.div(pow10(exponent.abs()))

  return positive_output.mul(sign)
}
