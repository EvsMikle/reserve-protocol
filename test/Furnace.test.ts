import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import { expect } from 'chai'
import { BigNumber, ContractFactory, Wallet } from 'ethers'
import hre, { ethers, waffle } from 'hardhat'
import { ZERO_ADDRESS } from '../common/constants'
import { bn } from '../common/numbers'
import {
  CTokenMock,
  ERC20Mock,
  FacadeP0,
  FurnaceP0,
  MainP0,
  TestIRToken,
  StaticATokenMock,
  USDCMock,
} from '../typechain'
import { advanceTime } from './utils/time'
import { Collateral, defaultFixture, IConfig } from './fixtures'
import { makeDecayFn } from './utils/rewards'

const createFixtureLoader = waffle.createFixtureLoader

describe('FurnaceP0 contract', () => {
  let owner: SignerWithAddress
  let addr1: SignerWithAddress
  let addr2: SignerWithAddress

  // Contracts
  let FurnaceFactory: ContractFactory
  let main: MainP0
  let furnace: FurnaceP0
  let rToken: TestIRToken
  let basket: Collateral[]
  let facade: FacadeP0

  // Config
  let config: IConfig

  // Tokens/Assets
  let token0: ERC20Mock
  let token1: ERC20Mock
  let token2: StaticATokenMock
  let token3: CTokenMock

  let collateral0: Collateral
  let collateral1: Collateral
  let collateral2: Collateral
  let collateral3: Collateral

  let initialBal: BigNumber

  let loadFixture: ReturnType<typeof createFixtureLoader>
  let wallet: Wallet

  before('create fixture loader', async () => {
    ;[wallet] = await (ethers as any).getSigners()
    loadFixture = createFixtureLoader([wallet])
  })

  beforeEach(async () => {
    ;[owner, addr1, addr2] = await ethers.getSigners()

    // Deploy fixture
    ;({ main, basket, rToken, furnace, config, facade } = await loadFixture(defaultFixture))

    // Setup issuance of RTokens for users
    initialBal = bn('100e18')

    // Get assets and tokens
    ;[collateral0, collateral1, collateral2, collateral3] = basket

    token0 = <ERC20Mock>await ethers.getContractAt('ERC20Mock', await collateral0.erc20())
    token1 = <USDCMock>await ethers.getContractAt('USDCMock', await collateral1.erc20())
    token2 = <StaticATokenMock>(
      await ethers.getContractAt('StaticATokenMock', await collateral2.erc20())
    )
    token3 = <CTokenMock>await ethers.getContractAt('CTokenMock', await collateral3.erc20())

    // Mint Tokens
    await token0.connect(owner).mint(addr1.address, initialBal)
    await token1.connect(owner).mint(addr1.address, initialBal)
    await token2.connect(owner).mint(addr1.address, initialBal)
    await token3.connect(owner).mint(addr1.address, initialBal)

    await token0.connect(owner).mint(addr2.address, initialBal)
    await token1.connect(owner).mint(addr2.address, initialBal)
    await token2.connect(owner).mint(addr2.address, initialBal)
    await token3.connect(owner).mint(addr2.address, initialBal)

    // Set Furnace Factory
    FurnaceFactory = await ethers.getContractFactory('FurnaceP0')
  })

  describe('Deployment', () => {
    it('Deployment should setup Furnace correctly', async () => {
      expect(await furnace.period()).to.equal(config.rewardPeriod)
      expect(await furnace.ratio()).to.equal(config.rewardRatio)
      expect(await furnace.lastPayout()).to.be.gt(0) // A timestamp is set
      expect(await furnace.main()).to.equal(main.address)
    })

    it('Deployment does not accept empty period', async () => {
      const newConfig = JSON.parse(JSON.stringify(config))
      newConfig.rewardPeriod = bn('0')
      const newFurnace = await FurnaceFactory.deploy()
      await expect(
        newFurnace.initComponent(main.address, {
          params: newConfig,
          components: {
            rToken: ZERO_ADDRESS,
            stRSR: ZERO_ADDRESS,
            assetRegistry: ZERO_ADDRESS,
            basketHandler: ZERO_ADDRESS,
            backingManager: ZERO_ADDRESS,
            distributor: ZERO_ADDRESS,
            rsrTrader: ZERO_ADDRESS,
            rTokenTrader: ZERO_ADDRESS,
            furnace: ZERO_ADDRESS,
            broker: ZERO_ADDRESS,
          },
          gnosis: ZERO_ADDRESS,
          assets: [ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS],
          rsr: ZERO_ADDRESS,
        })
      ).to.be.revertedWith('period cannot be zero')
    })
  })

  describe('Configuration / State', () => {
    it('Should allow to update period correctly if Owner and perform validations', async () => {
      // Setup a new value
      const newRewardPeriod: BigNumber = bn('100000')

      await expect(furnace.connect(owner).setPeriod(newRewardPeriod))
        .to.emit(furnace, 'PeriodSet')
        .withArgs(config.rewardPeriod, newRewardPeriod)

      expect(await furnace.period()).to.equal(newRewardPeriod)

      // Try to update again if not owner
      await expect(furnace.connect(addr1).setPeriod(bn('500'))).to.be.revertedWith(
        'Component: caller is not the owner'
      )

      // Cannot update with period zero
      await expect(furnace.connect(owner).setPeriod(bn('0'))).to.be.revertedWith(
        'period cannot be zero'
      )
    })

    it('Should allow to update ratio correctly if Owner', async () => {
      // Setup a new value
      const newRatio: BigNumber = bn('100000')

      await expect(furnace.connect(owner).setRatio(newRatio))
        .to.emit(furnace, 'RatioSet')
        .withArgs(config.rewardRatio, newRatio)

      expect(await furnace.ratio()).to.equal(newRatio)

      // Try to update again if not owner
      await expect(furnace.connect(addr1).setRatio(bn('0'))).to.be.revertedWith(
        'Component: caller is not the owner'
      )
    })
  })

  describe('Do Melt', () => {
    beforeEach(async () => {
      // Approvals for issuance
      await token0.connect(addr1).approve(rToken.address, initialBal)
      await token1.connect(addr1).approve(rToken.address, initialBal)
      await token2.connect(addr1).approve(rToken.address, initialBal)
      await token3.connect(addr1).approve(rToken.address, initialBal)

      await token0.connect(addr2).approve(rToken.address, initialBal)
      await token1.connect(addr2).approve(rToken.address, initialBal)
      await token2.connect(addr2).approve(rToken.address, initialBal)
      await token3.connect(addr2).approve(rToken.address, initialBal)

      // Issue tokens
      const issueAmount: BigNumber = bn('100e18')
      await rToken.connect(addr1).issue(issueAmount)
      await rToken.connect(addr2).issue(issueAmount)
    })

    it('Should not melt any funds in the initial block', async () => {
      const hndAmt: BigNumber = bn('2e18')

      // Set automine to false for multiple transactions in one block
      await hre.network.provider.send('evm_setAutomine', [false])

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Melt
      await furnace.connect(addr1).melt()

      // Mine block
      await hre.network.provider.send('evm_mine', [])

      // Check melt was not processed
      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Set automine to true again
      await hre.network.provider.send('evm_setAutomine', [true])
    })

    it('Should not melt if no funds available', async () => {
      // Set time period
      const period: number = 60 * 60 * 24 // 1 day
      await furnace.connect(owner).setPeriod(period)

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal)
      expect(await rToken.balanceOf(furnace.address)).to.equal(0)

      // Advance to the end to melt full amount
      await advanceTime(period + 1)

      // Melt
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      // Check nothing changed
      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal)
      expect(await rToken.balanceOf(furnace.address)).to.equal(0)
    })

    it('Should melt 0 for first period, even if funds available', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Advance one period
      await advanceTime(period + 1)

      // Melt
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      // Another call to melt should also have no impact
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)
    })

    it('Should allow melt - one period', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Advance to the end to melt full amount
      await advanceTime(period + 1)

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt = decayFn(hndAmt, 1) // 1 period

      // Melt
      await expect(furnace.connect(addr1).melt())
        .to.emit(rToken, 'Melted')
        .withArgs(hndAmt.sub(expAmt))

      // Another call to melt should have no impact
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(expAmt)
    })

    it('Should allow melt - two periods, all at once', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Advance to the end to melt full amount
      await advanceTime(2 * period + 1)

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt = decayFn(hndAmt, 2) // 2 periods

      await expect(furnace.melt()).to.emit(rToken, 'Melted').withArgs(hndAmt.sub(expAmt))

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(expAmt)
    })

    it('Should allow melt - two periods, one at a time', async () => {
      const hndAmt: BigNumber = bn('10e18')
      const period: number = 60 * 60 * 24 // 1 day

      // Set time period
      await furnace.connect(owner).setPeriod(period)

      // Transfer
      await rToken.connect(addr1).transfer(furnace.address, hndAmt)

      // Get past first noop melt
      await advanceTime(period + 1)
      await expect(furnace.connect(addr1).melt()).to.not.emit(rToken, 'Melted')

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(hndAmt)

      // Advance to the end to melt full amount
      await advanceTime(period + 1)

      const decayFn = makeDecayFn(await furnace.ratio())
      const expAmt1 = decayFn(hndAmt, 1) // 1 period

      // Melt
      await expect(furnace.connect(addr1).melt())
        .to.emit(rToken, 'Melted')
        .withArgs(hndAmt.sub(expAmt1))

      // Advance to the end to withdraw full amount
      await advanceTime(period + 1)

      const expAmt2 = decayFn(hndAmt, 2) // 2 periods

      // Melt
      await expect(furnace.connect(addr1).melt())
        .to.emit(rToken, 'Melted')
        .withArgs(bn(expAmt1).sub(expAmt2))

      expect(await rToken.balanceOf(addr1.address)).to.equal(initialBal.sub(hndAmt))
      expect(await rToken.balanceOf(furnace.address)).to.equal(expAmt2)
    })
  })
})
