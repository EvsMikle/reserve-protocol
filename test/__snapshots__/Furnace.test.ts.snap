// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, all at once 1`] = `82506`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, all at once 2`] = `99198`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 1`] = `82506`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 2`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 3`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 4`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 5`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 6`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 7`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 8`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 9`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 10`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - Many periods, one after the other 11`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - One period  1`] = `41059`;

exports[`FurnaceP1 contract Gas Reporting Melt - One period  2`] = `82506`;

exports[`FurnaceP1 contract Gas Reporting Melt - One period  3`] = `97032`;

exports[`FurnaceP1 contract Gas Reporting Melt - One period  4`] = `41059`;
